src/contract.rs

```diff
@@ -1239,93 +1409,69 @@ pub fn vote_proposal(
             val: "No tokens locked to perform voting on proposals".to_string(),
         });
     }
-
-    let vtokens = vtokens.unwrap();
-    // calculate voting power for the the proposal
-    let mut vote_power: u128 = 0;
-
-    for vtoken in vtokens {
-        vote_power += vtoken.vtoken.amount.u128();
+    if let Some(_vtokens) = vtokens {
+        // calculate voting power for the proposal
+        for vtoken in _vtokens {
+            vote_power += vtoken.vtoken.amount.u128();
+        }
     }
 
-    // Update proposal Vote for an app
-
-    let mut proposal_vote = PROPOSALVOTE
-        .load(deps.storage, (proposal_id, extended_pair))
-        .unwrap_or_default();
+    if vote_power == 0 {
+        return Err(ContractError::CustomError {
+            val: "No tokens locked to perform voting on proposals".to_string(),
+        });
+    }
 
-    // if already voted , update voting stats
+    //if already voted , decrease previous vote weight
     if has_voted {
         let prev_vote = VOTERSPROPOSAL.load(deps.storage, (info.sender.clone(), proposal_id))?;
-        let last_vote_weight = prev_vote.vote_weight;
-        let last_voted_pair = prev_vote.extended_pair;
-        if last_voted_pair == extended_pair {
-            proposal_vote -= Uint128::from(last_vote_weight);
-            proposal_vote += Uint128::from(vote_power);
-            proposal.total_voted_weight -= last_vote_weight;
-            proposal.total_voted_weight += vote_power;
-            PROPOSALVOTE.save(deps.storage, (proposal_id, extended_pair), &proposal_vote)?;
-            PROPOSAL.save(deps.storage, proposal_id, &proposal)?;
-            let app_id_param = app_id;
-            let extended_pair_param = extended_pair;
-            let vote = Vote {
-                app_id: app_id_param,
-                extended_pair: extended_pair_param,
-                vote_weight: vote_power,
-            };
-
-            VOTERSPROPOSAL.save(deps.storage, (info.sender.clone(), proposal_id), &vote)?;
-        } else {
-            let mut prev_proposal_vote = PROPOSALVOTE
-                .load(deps.storage, (proposal_id, last_voted_pair))
+        let last_vote_weight = prev_vote.votes;
+        for pair_vote in last_vote_weight {
```

Comment:
All previous votes are being removed even if they are not being updated.

Reply:
votersproposals is overridden later on in the function. prev vote is just used here to decrease individual pair vote weight. so it is expected